### PR TITLE
Faculty leave spec

### LINKED ISSUE #9: The FacultyLeave App should have a published API Spec
```
POST /leave
req. Employee, WorkAdmin, or HRAdmin

requet body:

201 response body:
401 response body:
403 response body:
422 response body:
500 response body:




GET /leave ? {query by employeeID (comma seperated), departmentName, timeframe}
req. Employee, WorkAdmin, or HRAdmin

200 response body:
400 response body:
500 response body:




GET /leave/{id}
req. Employee, WorkAdmin, or HRAdmin

200 response body:
400 response body:
404 response body:
500 response body:




PUT /leave/{id}
req. WorkAdmin, or HRAdmin

requet body:

204 response body:
401 response body:
403 response body:
404 response body:
422 response body:
500 response body:




PUT /leave/{id}/approve
req. WorkAdmin, or HRAdmin

204 response body:
401 response body:
403 response body:
404 response body:
500 response body:




PUT /leave/{id}/reject
req. WorkAdmin, or HRAdmin

204 response body:
401 response body:
403 response body:
404 response body:
500 response body:




PUT /leave/{id}/cancel
req. Employee, WorkAdmin, or HRAdmin

204 response body:
401 response body:
403 response body:
404 response body:
500 response body:




POST /leave/{id}/correction
req. Employee

requet body:

201 response body:
401 response body:
403 response body:
422 response body:
500 response body:




GET /employees/{id}
req. Employee, WorkAdmin, or HRAdmin

response body:

200 response body:
401 response body:
403 response body:
404 response body:
500 response body:




GET /employees/{id}/leave
req. Employee, WorkAdmin, or HRAdmin

200 response body:
401 response body:
403 response body:
404 response body:
500 response body:




GET /employees/{id}/history {query by timeframe}
req. Employee. WorkAdmin. or HRAdmin.

200 response body:
400 response body:
401 response body:
403 response body:
404 response body:
500 response body:




GET /employees/{id}/balance
req. Employee, WorkAdmin, or HRAdmin.

200 response body:
400 response body:
401 response body:
403 response body:
404 response body:
500 response body:




GET /employees/{id}/settings
req. Employee, WorkAdmin, or HRAdmin

200 response body:
400 response body:
401 response body:
403 response body:
404 response body:
500 response body:




PUT /employees/{id}/settings
req. Employee, WorkAdmin, or HRAdmin

requet body:

204 response body:
400 response body:
401 response body:
403 response body:
404 response body:
422 response body:
500 response body:
```

https://editor.swagger.io/

### PR DESCRIPTION
**Pull request by jeremythuff**
Adds FacultyLeave specefication

Resolves #9

### CODE REVIEW

File: openapi-docs/faculty-leave.yaml [hunk 1]
```diff
@@ -0,0 +1,662 @@
...
+  license:
+    name: Apache 2.0
+    url: http://www.apache.org/licenses/LICENSE-2.0.html
+  version: 1.0.0
+externalDocs:
+  description: Find out more about Swagger
+  url: http://swagger.io
+servers:
+- url: https://github.com/TAMULib/FacultyLeave
+- url: http://github.com/TAMULib/FacultyLeave
```
Comment (ghost) [inline openapi-docs/faculty-leave.yaml]: Was it intentional to have two duplicate URLs with different protocols?

File: openapi-docs/faculty-leave.yaml [hunk 2]
```diff
@@ -0,0 +1,662 @@
...
+      description: Get a list of requests for leave.
+      operationId: getLeave
+      security:
+        - spring_session: [
+          Employee, WorkAdmin, or HRAdmin
+        ]
+      parameters:
+      - name: ids
+        in: query
+        description: Comma seperated string of ids to return
```
Comment (ghost) [inline openapi-docs/faculty-leave.yaml]: This does not read as I would expect. It reads as if the ids would be returned.

Suggest

`Comma separated string of request ids`

or

`Comma separated string of leave request ids`

or something similar.

File: openapi-docs/faculty-leave.yaml [hunk 3]
```diff
@@ -0,0 +1,662 @@
...
+        500:
+          description: Internal Server Error
+          content: {}
+      x-codegen-request-body-name: body
+  /leave/{id}:
+    get:
+      tags:
+      - leave
+      summary: Get a request for leave by id
+      description: Get a list of requests for leave.
```
Comment (ghost) [inline openapi-docs/faculty-leave.yaml]: Is this accurate? Is it returning an individual leave request or a list?

File: openapi-docs/faculty-leave.yaml [hunk 4]
```diff
@@ -0,0 +1,662 @@
...
+          content: {}
+        500:
+          description: Internal Server Error
+          content: {}
+  /leave/{id}/reject:
+    put:
+      tags:
+      - leave
+      summary: Modify a leave request with rejection.
+      description: Modify a leave request with approval.
```
Comment (ghost) [inline openapi-docs/faculty-leave.yaml]: Should be `with rejection`

File: openapi-docs/faculty-leave.yaml [hunk 5]
```diff
@@ -0,0 +1,662 @@
...
+          description: Leave request not found
+          content: {}
+        500:
+          description: Internal Server Error
+          content: {}
+  /leave/{id}/correction:
+    post:
+      tags:
+      - leave
+      summary: Add a new corection for an existing request for leave.
```
Comment (ghost) [inline openapi-docs/faculty-leave.yaml]: Shouldn't this be `Request a correction`?

File: openapi-docs/faculty-leave.yaml [hunk 6]
```diff
@@ -0,0 +1,662 @@
...
+      requestBody:
+        description: PastLeaveChangeRequest object describing the requested changes.
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/PastLeaveChangeRequest'
+        required: true
+      responses:
+        201:
+          description: Leave Request Created
```
Comment (ghost) [inline openapi-docs/faculty-leave.yaml]: These responses should correspond to the request for correction.

File: openapi-docs/faculty-leave.yaml [hunk 7]
```diff
@@ -0,0 +1,662 @@
...
+          description: Employee Not Found
+          content: {}
+        500:
+          description: Internal Server Error
+          content: {}
+  /employees/{uin}/leave:
+    get:
+      tags:
+      - employees
+      summary: Retrieves an employee's requested leave by UIN
```
Comment (ghost) [inline openapi-docs/faculty-leave.yaml]: Should this be to retrieve a list of request leave for an employee?

File: openapi-docs/faculty-leave.yaml [hunk 8]
```diff
@@ -0,0 +1,662 @@
...
+      parameters:
+      - name: uin
+        in: path
+        required: true
+        schema:
+          type: string
+          format: uuid
+      responses:
+        200:
+          description: Employee Leave History Found
```
Comment (ghost) [inline openapi-docs/faculty-leave.yaml]: This should be the description of employee leave balance.

**Review by ghost — CHANGES_REQUESTED** [review]
It seems as we will be iterating on this as the application is developed. The spring security values will likely have to be updated to match in code. I believe we will also want example requests and responses to make it a more useful API in the specifications UI.

File: openapi-docs/faculty-leave.yaml (hunk 1 above)
Comment (jeremythuff) [reply openapi-docs/faculty-leave.yaml]: These are placeholders until we get an actual server up

File: openapi-docs/faculty-leave.yaml [hunk 9]
```diff
@@ -0,0 +1,662 @@
+openapi: 3.0.1
+info:
+  title: FacultyLeave API Specefication
+  description: The API Specefication for Texas A&M Libraries FacultyLeave Application
```
Comment (ghost) [inline openapi-docs/faculty-leave.yaml]: Typo

Specefication => Specification

File: openapi-docs/faculty-leave.yaml [hunk 10]
```diff
@@ -0,0 +1,662 @@
...
+  description: Find out more about Swagger
+  url: http://swagger.io
+servers:
+- url: https://github.com/TAMULib/FacultyLeave
+paths:
+  /leave:
+    get:
+      tags:
+      - leave
+      summary: Get a list of requests for leave.
```
Comment (ghost) [inline openapi-docs/faculty-leave.yaml]: Should we be consistent with punctiation?

**Review by ghost — CHANGES_REQUESTED** [review]
Sorry, missed these two on the first pass.

File: openapi-docs/faculty-leave.yaml [hunk 11]
```diff
@@ -1,7 +1,7 @@
 openapi: 3.0.1
 info:
   title: FacultyLeave API Specefication
```
Comment (ghost) [inline openapi-docs/faculty-leave.yaml]: Same typo.

**Review by ghost — APPROVED** [review]

**Pull request merged**